cip/1.accepted/CIP2017-06-18-multiple-graphs.adoc

```diff
@@ -0,0 +1,944 @@
+
+== Proposal: Multiple Graphs Syntax
+
+This CIP first proposes new syntactical concepts before proceeding to add new and extend existing clauses.
+
+=== Named graphs
+
+This CIP introduces the notion of multiple named graphs.
+
+A graph is referenced in the current query context via its name.
+Graph names `<graph-name>` use the same syntax as variable names.
+It is an error to use simultaneously the same name for both a regular variable and a graph.
+
+==== Graph references and aliases
+
+An explicit reference to a `<graph-ref>` is just the `<graph-name>` of the graph.
+A `<graph-alias>` is a `<graph-ref>` that is optionally followed by `AS <new-graph-name>`.
+A `<graph-alias-list>` is a comma-separated list of `<graph-alias>`.
+
+==== Graph definitions
+
+This CIP defines the notion of a graph definition `<graph-def>` as a means by which to introduce additional named graphs.
+This CIP proposes the following kinds of graph definitions:
+
+* `NEW GRAPH <new-graph-name> [AT <graph-url]`: A newly created empty graph that is to be made available at the provided `<graph-url>`
+* `COPY GRAPH <new-graph-name> FROM <graph-ref> [TO <graph-url>]`: A newly created copy of the given graph `<graph-ref>` that is to be made available as `<new-graph-name>` at the provided `<graph-url>`
+* `GRAPH <new-graph-name> AT <graph-url>`: The graph at the provided `<graph-url>` that is to be made available as `<new-graph-name>`
+* `GRAPH <graph-alias>` (i.e. `GRAPH <graph-ref> [AS <new-graph-name>]`): A new alias for an existing graph
+* `SOURCE GRAPH AS <new-graph-name>`: An alias for the current _source graph_
+* `TARGET GRAPH AS <new-graph-name>`: An alias for the current _target graph_
+* `DEFAULT GRAPH AS <new-graph-name>`: An alias for the current _default graph_
+
+Note: The syntax follows the simple rule that if a graph definition aliases a definitely already existing graph, then the alias name of that graph is given last using `AS`.
+Otherwise, the name is given right after the leading keywords of the graph definition.
+
+Note: The exact shape and form of graph URLs `<graph-url>` lies outside the scope of this CIP.
+However, this CIP proposes that a `<graph-url>` must always be given as either a string literal or a query parameter.
+This allows parameterization of queries by controlling which graphs from which graph URLs they should use.
+
+==== Graph specifiers
+
+This CIP defines the notion of a graph specifier `<graph-spec>` to be either a `<graph-def>` or  `DEFAULT GRAPH` for referring to the default graph without naming it.
+
+=== Introducing multiple graphs
+
+As a first language addition, this CIP proposes syntax for introducing graphs into the current query context:
+
+[source, cypher]
+----
+FROM < graph-spec >
+INTO < graph-spec >
+----
+
+==== FROM clause: Change the source and the target graph
+
+This CIP proposes a new `FROM` clause to change both the source and the target graph of the current query context as described.
```

Comment:
I'm not convinced that having `FROM` change _both_ source and target is the best choice, but I'm also not sure how to argue in either direction. It feels easier to grasp if `FROM` is tied to source and `INTO` tied to target.